### ISSUE TITLE
[feature request] Subset of eigenvalues/eigenvectors 

### ISSUE THREAD

**Issue by LucaMoschella** [pytorch/pytorch#24185]
## 🚀 Feature
Computation of a subset of eigenvalues and eigenvectors

Inspired by [MATLAB eigs](https://it.mathworks.com/help/matlab/ref/eigs.html#bu2_q3e-sigma):

    d = torch.eigs(A, B, k, sigma)

Solves the generalized eigenvalue problem `A * V = d * B * V `

- `k`: Number of eigenvalues to compute, specified as a positive scalar integer.
- `sigma`: The eigenvalues closest to the number sigma.

## Motivation
I need `k` eigenvalues of a `n * n` matrix, currently it's an intractable problem in PyTorch.

e.g. with `k=30` and `n=500`

MATLAB:
```matlab
[phi, e] = eigs(S, L, 30, -1e-5);
% Elapsed time is 0.074229 seconds.
```

NumPy:
```python
e, phi = scipy.sparse.linalg.eigs(S, 30, L, sigma=-1e-5)
# Elapsed time is 2.5 seconds.
```

PyTorch:
```python
e, phi = torch.symeig(C, eigenvectors=True)

# cpu
# Elapsed time is 5 seconds plus the time for the cholesky decomposition. 

# cuda
# Elapsed time is 2 seconds plus the time for the cholesky decomposition. 
```

## Pitch
I want to be able to compute the subset of eigenvalues/eigenvectors of a matrix that I need, without computing all the others.

## Alternatives
Computing all the eigenvalues/eigenvectors and then selecting the relevant ones.

## Usecase 
The [Laplace-Beltrami](https://en.wikipedia.org/wiki/Laplace%E2%80%93Beltrami_operator) operator is a key tool in geometry processing, and its eigenvalues/eigenvectors are an optimal basis to represent smooth functions on surfaces.
For example, it's common to project functions into the basis of LB eigenvectors, truncating the base to the first k-elements. This is equivalent to the low-pass filtering of the Fourier series:

![spec8](https://user-images.githubusercontent.com/11019190/62874586-d2d11900-bd21-11e9-8bbd-cb646af36568.png)

Moreover, the associated eigenvalues encode [properties of the surface](https://arxiv.org/abs/1811.11465), e.g. when sorted they lie on a line and the slope of this line is proportional to the surface's area.

If only `k` elements of the spectrum will be used, it is extremely wasteful computing all the eigenvalues/eigenvectors of the Laplace-Beltrami operator, since when discretized their number is equal to the number of vertices in the mesh, yielding an intractable problem in practice.

**Comment by soumith** [comment]
cc: @vishwakftw

**Comment by vishwakftw** [comment]
Thank you opening this feature request, @LucaMoschella.

The closest related LAPACK and MAGMA routines to do this are `syevx` and `syevdx_gpu` respectively.

However, these routines do either one of the following:
- compute the eigenvalues and corresponding eigenvectors within a half open interval `(vl, vu]`.
- compute the mth largest to nth largest eigenvalues and their corresponding eigenvectors.

I'm not sure how MATLAB performs the computation, but SciPy uses ARPACK routines for the exact functionality that you have described in your feature request, and the aforementioned LAPACK and MAGMA routines deviate from your requirements a bit. For instance, you wouldn't be able to compute eigenvalues close to a given `sigma`.

**Comment by LucaMoschella** [comment]
Thank you for your kind reply @vishwakftw 

I think the main feature that's missing is the ability to compute a subset of k eigenvectors/eigenvalues. 
Would it be possible to compute the `k` smallest (or largest) eigenvalues/eigenvectors?

I described the functionality that uses `sigma` only because it is the one that I'm using in MATLAB, I'm sorry if I wasn't clear enough in the previous post.

**Comment by vadimkantorov** [comment]
This is also related (top singular vectors): https://github.com/pytorch/pytorch/issues/8049

**Comment by leotrs** [comment]
Is there any update on this? It seems surprising that PyTorch doesn't support easy computation of, say, the leading eigenvalue of a matrix. These have myriad applications and can be done much more efficiently than computing all of the eigenvalues.

**Comment by lyndond** [comment]
@leotrs methods to accomplish this and more are actively being worked on in #29488

**Comment by simbilod** [comment]
In addition to the above, a recent [arxiv submission](https://arxiv.org/abs/2001.04121) seems to tackle this ([repo here](https://github.com/buwantaiji/DominantSparseEigenAD)). I have not evaluated how well it works though.

**Comment by MyEvolution** [comment]
There is still no update on this problem???

**Comment by vadimkantorov** [comment]
@lezcano @nikitaved Btw does PyTorch have any sparse eigensolver/generalized eigensolver? (for discrete local laplacian problems)

I found that SciPy has: https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.eigsh.html

**Comment by lezcano** [comment]
Yeah, a different issue should be better. Link this one in that one.

**Comment by vadimkantorov** [comment]
Btw the OP in this thread also mentions a sparse .eigsh

**Comment by vadimkantorov** [comment]
@lezcano created https://github.com/pytorch/pytorch/issues/108977 :)

**Comment by vadimkantorov** [comment]
And also found: https://github.com/pytorch/pytorch/issues/58828
